storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3574,6 +3589,294 @@ dberr_t fsp_sys_tablespace_validate()
+@param page_ranges  store the range of pages exist in this segment */
+static void fseg_get_pages(const fseg_header_t *header, mtr_t *mtr,
+                           range_set *page_ranges)
+{
+  buf_block_t *iblock;
+  fseg_inode_t *inode= fseg_inode_try_get(header, 0, 0, mtr, &iblock);
+  const uint16_t ioffset= uint16_t(inode - iblock->page.frame);
+  fseg_get_n_frag_pages(inode, page_ranges);
+
+  flst_get_page_range(
+    iblock, uint16_t(ioffset + FSEG_NOT_FULL), mtr, page_ranges);
+
+  flst_get_page_range(
+    iblock, uint16_t(ioffset + FSEG_FULL), mtr, page_ranges);
+
+  flst_get_page_range(
+    iblock, uint16_t(ioffset + FSEG_FREE), mtr, page_ranges);
+
+  return;
+}
+
+#ifdef UNIV_DEBUG
+/** Print the system tablespace page type information
+@return error code */
+static
+dberr_t fsp_system_print_pages()
+{
+  fprintf(stderr, "Print the pages in the system tablespace\n\n\t");
+  fil_space_t *space= fil_system.sys_space;
+  uint32_t n_words= 0;
+  dberr_t err= DB_SUCCESS;
+  mtr_t mtr;
+  mtr.start();
+  for (uint32_t i= 0; i < space->free_limit; i++)
+  {
+    if (n_words == 16)
+    {
+      fprintf(stderr, "\n\t");
+      n_words= 0;
+    }
+    n_words++;
+    if (i >= buf_dblwr.get_block1().page_no()
+        && i <= buf_dblwr.get_block2().page_no() + FSP_EXTENT_SIZE - 1)
+    {
+      fprintf(stderr, "D");
+      continue;
+    }
+    buf_block_t *block= buf_page_get_gen(
+      page_id_t(space->id, i), 0, RW_S_LATCH, nullptr,
+      BUF_GET_POSSIBLY_FREED, &mtr, &err);
+    if (!block || block->page.state() <= buf_page_t::FREED)
+    {
+      fprintf(stderr, "F");
+      err= DB_SUCCESS;
+      continue;
+    }
+    ulint type= fil_page_get_type(block->page.frame);
+    switch (type) {
+    case FIL_PAGE_INDEX:
+      fprintf(stderr, "I");
+      break;
+    case FIL_PAGE_UNDO_LOG:
+      fprintf(stderr, "U");
+      break;
+    case FIL_PAGE_INODE:
+      fprintf(stderr, "@");
+      break;
+    case FIL_PAGE_IBUF_FREE_LIST:
+      fprintf(stderr, "$");
+      break;
+    case FIL_PAGE_TYPE_ALLOCATED:
+      fprintf(stderr, "#");
+      break;
+    case FIL_PAGE_IBUF_BITMAP:
+      fprintf(stderr, "&");
+      break;
+    case FIL_PAGE_TYPE_SYS:
+      fprintf(stderr, "S");
+      break;
+    case FIL_PAGE_TYPE_TRX_SYS:
+      fprintf(stderr, "T");
+      break;
+    case FIL_PAGE_TYPE_FSP_HDR:
+      fprintf(stderr, "H");
+      break;
+    case FIL_PAGE_TYPE_XDES:
+      fprintf(stderr, "X");
+      break;
+    case FIL_PAGE_TYPE_BLOB:
+    case FIL_PAGE_TYPE_ZBLOB:
+    case FIL_PAGE_TYPE_ZBLOB2:
+      fprintf(stderr, "B");
+      break;
+    case FIL_PAGE_PAGE_COMPRESSED:
+    case FIL_PAGE_PAGE_COMPRESSED_ENCRYPTED:
+      fprintf(stderr, "Z");
+      break;
+    default:
+      ut_ad(0);
+    }
+  }
+  mtr.commit();
+  fprintf(stderr, "\nNote the letter description of the pages\n\t");
+  fprintf(stderr, "I - Index node page\n\t"
+                  "U - Undo log page\n\t"
+                  "Z - Compressed page\n\t"
```

Comment:
There is nothing formatted in the output; therefore we could as well use `fputs` or `putc`.

I was hoping that we would display something like `'A' + ((index_id - 10) % ('~' - 'A' + 1))` for pages that look like index pages. That would help identify them. This would leave the ASCII range `' ' to '?'` (0x20 to 0x3f) available for displaying other leaked pages.

Compressed pages can’t exist in the system tablespace, neither `page_compressed` nor `ROW_FORMAT=COMPRESSED`.